File: 2021/src/main/java/de/jfriemel/aoc/days/Day01.java
package de.jfriemel.aoc.days;

import de.jfriemel.aoc.Day;

import java.util.List;

public class Day01 implements Day {

    @Override
    public String part1(final List<String> input) {
        int previous = Integer.MAX_VALUE;
        int current;
        int result = 0;
        for (final String value : input) {
            current = Integer.parseInt(value);
            if (current > previous) {
                result++;
            }
            previous = current;
        }
        return Integer.toString(result);
    }

    @Override
    public String part2(final List<String> input) {
        int a = Integer.parseInt(input.get(0));
        int b = Integer.parseInt(input.get(1));
        int c;
        int previous = Integer.MAX_VALUE;
        int current;
        int result = 0;
        for (int i = 2; i < input.size(); i++) {
            c = Integer.parseInt(input.get(i));
            current = a + b + c;
            if (current > previous) {
                result++;
            }
            a = b;
            b = c;
            previous = current;
        }
        return Integer.toString(result);
    }

}

File: 2023/src/main/kotlin/de/jfriemel/aoc/days/Day01.kt
package de.jfriemel.aoc.days

import de.jfriemel.aoc.Day

object Day01 : Day {
    override fun part1(input: List<String>): String {
        return input.fold(0) { acc, entry ->
            val digits = entry.filter { it.isDigit() }
            acc + "${digits.first()}${digits.last()}".toInt()
        }.toString()
    }

    override fun part2(input: List<String>): String {
        return input.fold(0) { acc, entry ->
            val fullEntry = entry
                .replace("one", "one1one")
                .replace("two", "two2two")
                .replace("three", "three3three")
                .replace("four", "four4four")
                .replace("five", "five5five")
                .replace("six", "six6six")
                .replace("seven", "seven7seven")
                .replace("eight", "eight8eight")
                .replace("nine", "nine9nine")
            val digits = fullEntry.filter { it.isDigit() }
            acc + "${digits.first()}${digits.last()}".toInt()
        }.toString()
    }
}
